### PR TITLE
Add `*pty` API

### LINKED ISSUE #3212: Add support for `forkpty` in RzSubprocess
RzSubprocess API does not have a way to use `forkpty` or any of the `*pty` APIs at all, making it hard to use in scenarios where a new TTY/PTY may be needed.

### PR DESCRIPTION
**Pull request by DMaroo**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Add API to access `*pty` functions (`openpty`, `login_tty` and `forkpty`)

**Test plan**

Add unit tests, green CI

**Closing issues**

Closes #3212

### CODE REVIEW

**Comment by DMaroo** [comment]
@ret2libc Does the current approach look good (having a `RzPty` struct, ignore the signatures for `forkpty` and `login_tty`)?

I am planning to integrate `RzSubprocess` with `forkpty`.

**Review by ret2libc — COMMENTED** [review]
Hey @DMaroo thanks a lot!

A couple of comments:
- in general I prefer to have an API always available and just return an error in the not-implemented version. This way users of the API don't have to do other #ifdef to use the API.
- I was thinking more about integrating the pty functions like forkpty with the RzSubprocess so that a user could just use existing rz_subprocess_* APIs but with few options to choose a new PTY instead of a regular fork. Do you see what I mean? Like, I'd like to do something similar to `test_subprocess.c:test_interactive()` but with a new PTY for stdin.


Let me know what you think, and thanks again for working on this!

**Comment by DMaroo** [comment]
Yup. Understood. I'll change the `RzSubprocessOpt` to also have a flag about whether the forking should be done by `forkpty` or by `fork`.

**Review by ret2libc — CHANGES_REQUESTED** [review]
Please add also a test if possible, to ensure that the API makes sense and that it really creates a separate PTY.

Looking forward to this!! With this we can then implement well few RzRun features.

File: librz/util/sys.c [hunk 1]
```diff
@@ -1681,6 +1693,33 @@ RZ_API int rz_sys_fork(void) {
 }
 #endif
 
+RZ_API pid_t rz_sys_forkpty(int *amaster, char *name, const void /* struct termios */ *termp, const void /* struct winsize */ *winp) {
+#if HAVE_OPENPTY && HAVE_FORKPTY && HAVE_LOGIN_TTY
+	return forkpty(amaster, name, (const struct termios *)termp, (const struct winsize *)winp);
+#else
+	RZ_LOG_ERROR("\"forkpty\" not found\n");
```
Comment (XVilka) [inline librz/util/sys.c]: `forkpty() not found`, do that's clear it's a function.

File: librz/util/sys.c [hunk 2]
```diff
@@ -1681,6 +1693,33 @@ RZ_API int rz_sys_fork(void) {
...
+	return -1;
+#endif
+}
+
+RZ_API int rz_sys_openpty(int *amaster, int *aslave, char *name, const void /* struct termios */ *termp, const void /* struct winsize */ *winp) {
+#if HAVE_OPENPTY && HAVE_FORKPTY && HAVE_LOGIN_TTY
+	return openpty(amaster, aslave, name, (const struct termios *)termp, (const struct winsize *)winp);
+#else
+	RZ_LOG_ERROR("\"openpty\" not found\n");
```
Comment (XVilka) [inline librz/util/sys.c]: Same

File: librz/util/sys.c [hunk 3]
```diff
@@ -1681,6 +1693,33 @@ RZ_API int rz_sys_fork(void) {
...
+	return -1;
+#endif
+}
+
+RZ_API int rz_sys_login_tty(int fd) {
+#if HAVE_OPENPTY && HAVE_FORKPTY && HAVE_LOGIN_TTY
+	return login_tty(fd);
+#else
+	RZ_LOG_ERROR("\"login_tty\" not found\n");
```
Comment (XVilka) [inline librz/util/sys.c]: Same

File: librz/util/subprocess.c [hunk 4]
```diff
@@ -1355,6 +1375,54 @@ RZ_API RzSubprocess *rz_subprocess_start(
 		.stdin_pipe = RZ_SUBPROCESS_PIPE_CREATE,
 		.stdout_pipe = RZ_SUBPROCESS_PIPE_CREATE,
 		.stderr_pipe = RZ_SUBPROCESS_PIPE_CREATE,
+		.fork_mode = RZ_SUBPROCESS_FORK,
+		.pty = NULL
+	};
+	return rz_subprocess_start_opt(&opt);
+}
+
+RZ_API RzPty *rz_subprocess_openpty(RZ_NULLABLE RZ_BORROW char *slave_name, RZ_NULLABLE const void /* struct termios */ *term_params, RZ_NULLABLE const void /* struct winsize */ *win_params) {
```
Comment (XVilka) [inline librz/util/subprocess.c]: Add ownership mark

File: librz/include/rz_util/rz_subprocess.h [hunk 5]
```diff
@@ -95,4 +111,8 @@ RZ_API RzStrBuf *rz_subprocess_stdout_readline(RzSubprocess *proc, ut64 timeout_
 RZ_API RzSubprocessOutput *rz_subprocess_drain(RzSubprocess *proc);
 RZ_API void rz_subprocess_output_free(RzSubprocessOutput *out);
 
-#endif
\ No newline at end of file
+RZ_API RzPty *rz_subprocess_openpty(RZ_NULLABLE RZ_BORROW char *slave_name, RZ_NULLABLE const void /* struct termios */ *term_params, RZ_NULLABLE const void /* struct winsize */ *win_params);
```
Comment (XVilka) [inline librz/include/rz_util/rz_subprocess.h]: Ownership mark

File: librz/util/subprocess.c [hunk 6]
```diff
@@ -1355,6 +1375,54 @@ RZ_API RzSubprocess *rz_subprocess_start(
...
+	int ret = rz_sys_login_tty(pty->slave_fd);
+	if (ret == -1) {
+		perror("login_tty");
+		return false;
+	}
+
+	return true;
+}
+
+RZ_API RzSubprocess *rz_subprocess_forkpty(const char *file, const char *args[], size_t args_size,
```
Comment (XVilka) [inline librz/util/subprocess.c]: Ownership mark

**Comment by XVilka** [comment]
Resend this from `dist-`, please, to trigger all possible builds.

**Comment by XVilka** [comment]
It fails on MacOS:
```
Exception in callback TestHarness._run_tests.<locals>.test_done(<Task finishe...t permitted')>) at /Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py:1922
handle: <Handle TestHarness._run_tests.<locals>.test_done(<Task finishe...t permitted')>) at /Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py:1922>
Traceback (most recent call last):
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1306, in wait
    await complete_all(self.all_futures, timeout=test.timeout)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1206, in complete_all
    check_futures(futures)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1194, in check_futures
    raise asyncio.TimeoutError
TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):

  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/events.py", line 80, in _run
Ok:                 98  
    self._context.run(self._callback, *self._args)
Expected Fail:      0   
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1924, in test_done
Fail:               0   
    f.result()
Unexpected Pass:    0   
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1916, in run_test
Skipped:            0   
    res = await test.run(self)
Timeout:            0   
          ^^^^^^^^^^^^^^^^^^^^

  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1419, in run
Full log written to /Users/runner/work/rizin/rizin/build/meson-logs/testlog.txt
    await self._run_cmd(harness, cmd)

  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1486, in _run_cmd
ERROR: Unhandled python OSError. This is probably not a Meson bug, but an issue with your build environment.
    await p.wait(self.runobj)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1308, in wait
    test.additional_error += await self._kill() or ''
                             ^^^^^^^^^^^^^^^^^^
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1273, in _kill
    os.killpg(p.pid, signal.SIGKILL)
PermissionError: [Errno 1] Operation not permitted
Traceback (most recent call last):
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1306, in wait
    await complete_all(self.all_futures, timeout=test.timeout)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1206, in complete_all
    check_futures(futures)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1194, in check_futures
    raise asyncio.TimeoutError
TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 2073, in run
    return th.doit()
           ^^^^^^^^^
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1754, in doit
    self.run_tests(runners)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1893, in run_tests
    loop.run_until_complete(self._run_tests(runners))
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 650, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1993, in _run_tests
    await complete_all(futures)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1202, in complete_all
    check_futures(done)
  File "/Users/runner/Library/Python/3.11/lib/python/site-packages/mesonbuild/mtest.py", line 1192, in check_futures
    f.result()
PermissionError: [Errno 1] Operation not permitted
Error: Process completed with exit code 1.
```

**Comment by DMaroo** [comment]
Closing in favour of #3221.

**Pull request closed without merging**